sessiondata: introduce sessiondata.Stack and thread to EvalContext and connExecutor

otan opened the pull request:
See individual commits for details. This one is needed ASAP to prevent merge conflicts.
Part 1 of a series of PRs to merge SET LOCAL.
Refs: #32562

Release justification: low risk, high pri change.

otan committed:
sessiondata: introduce sessiondata.Stack

This stack keeps a track of SessionData objects for use in
transaction-scoped variables.

Release justification: low risk, high pri change

Release note: None

otan committed:
sql: replace sessionData in connExecutor with sessiondata.Stack

A mechanical change, the SessionData on the connExecutor now is replaced
with the sessiondata.Stack object.

Release justification: low risk, high pri change

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69338)
<!-- Reviewable:end -->

otan committed:
sql: thread SessionDataStack to EvalContext

Replace SessionData with sessiondata.Stack in EvalContext, and replace
every caller of SessionData with a supplied SessionData() method, which
fetches the top of the stack. This, by default, ensures sessions are
using the "local" SessionData to do their work.

Note as of this commit, there is only ever one element in the stack.

Release justification: low risk, high pri change

Release note: None

rafiss reviewed (approved):
this looks good to me! i had some nits/questions but in the interest of saving you possible merge conflicts, i'll merge this now and we can discuss them tonight when we're both online

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69338#-:-MhyL1PdEy_9HoVmYVJO:b-5q00hz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nihalpednekar and @otan)

___
*[pkg/sql/conn_executor.go, line 1480 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69338#-MhyKcxf2Nl9s4QsWb9l:-MhyKcxf2Nl9s4QsWb9m:b39dtmo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d71dd3245d1eeb2557c79cd158752c7dfd48dccc/pkg/sql/conn_executor.go#L1480)):*
> ```Go
> }
> 
> // sessionData returns the top SessionData on the executor.
> ```

nit: "the top SessionData in the executor's sesstionDataStack"

could also mention again here that it's used so we can have transaction-scoped session variable changes
___
*[pkg/sql/sessiondata/session_data.go, line 199 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69338#-MhyISCYBCk2YV0qpgtB:-MhyISCYBCk2YV0qpgtC:b9vpftn) ([raw file](https://github.com/cockroachdb/cockroach/blob/38eeb53e850c320335eb0f44d836a7f7fa81fb21/pkg/sql/sessiondata/session_data.go#L199)):*
> ```Go
> 
> // NewStack creates a new tack.
> func NewStack(firstElem *SessionData) *Stack {
> ```

nit: might be convenient to add another function that takes no parameters and creates a stack with an empty sessiondata. seems like a lot of tests want that

also it seems that no one outside of the `sql` package should ever need to initialize a `SessionData` struct. is there a way to enforce that?
___
*[pkg/sql/sessiondata/session_data.go, line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69338#-MhyGlh1BWpdoX0toCHj:-MhyGlh1BWpdoX0toCHk:b-qnpbdn) ([raw file](https://github.com/cockroachdb/cockroach/blob/38eeb53e850c320335eb0f44d836a7f7fa81fb21/pkg/sql/sessiondata/session_data.go#L206)):*
> ```Go
> func (s *Stack) Top() *SessionData {
> 	if len(s.stack) == 0 {
> 		return nil
> ```

should this function return an AssertionError here if the stack is empty?
___
*[pkg/sql/sessiondata/session_data.go, line 221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69338#-MhyK0jC-h3XHXpwjE_j:-MhyK0jC-h3XHXpwjE_k:b-q0qalv) ([raw file](https://github.com/cockroachdb/cockroach/blob/38eeb53e850c320335eb0f44d836a7f7fa81fb21/pkg/sql/sessiondata/session_data.go#L221)):*
> ```Go
> 		return errors.AssertionFailedf("there must always be at least one element in the SessionData stack")
> 	}
> 	s.stack = s.stack[:len(s.stack)-1]
> ```

just checking -- does the pointer that just got popped off get garbage collected? (i forget how it works, but asking because i was working on something else where a pointer into a backing array was causing memory issues though maybe that was because there were strings involved)


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69338)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nihalpednekar, @otan, and @rafiss)

___
*[pkg/sql/sessiondata/session_data.go, line 221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69338#-MhyK0jC-h3XHXpwjE_j:-MhyWb9uB8_95k0HYUoB:b9ex0ox) ([raw file](https://github.com/cockroachdb/cockroach/blob/38eeb53e850c320335eb0f44d836a7f7fa81fb21/pkg/sql/sessiondata/session_data.go#L221)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

just checking -- does the pointer that just got popped off get garbage collected? (i forget how it works, but asking because i was working on something else where a pointer into a backing array was causing memory issues though maybe that was because there were strings involved)
</blockquote></details>

we should probably assign the element to nil to avoid leaking it. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69338)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nihalpednekar, @otan, and @rafiss)

___
*[pkg/sql/sessiondata/session_data.go, line 221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69338#-MhyK0jC-h3XHXpwjE_j:-MhyWfJs-Yv-VhmZpHfN:b-cfvvr4) ([raw file](https://github.com/cockroachdb/cockroach/blob/38eeb53e850c320335eb0f44d836a7f7fa81fb21/pkg/sql/sessiondata/session_data.go#L221)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

we should probably assign the element to nil to avoid leaking it. 
</blockquote></details>

like:
```
s.stack[len(s.stack)-1] = nil
s.stack = s.stack[:len(s.stack)-1]
```


<!-- Sent from Reviewable.io -->
